lang/src/org/partiql/lang/planner/transforms/AstToLogicalVisitorTransform.kt

```diff
@@ -326,6 +365,55 @@ internal class AstToLogicalVisitorTransform(
                 }
             )
         }
+
+    override fun transformExprLit(node: PartiqlAst.Expr.Lit): PartiqlLogical.Expr =
+        (super.transformExprLit(node) as PartiqlLogical.Expr.Lit).apply { // Validation
+            if (value is IntElement && value.integerSize == IntElementSize.BIG_INTEGER) {
+                throw EvaluationException(
+                    message = "Int overflow or underflow at compile time",
+                    errorCode = ErrorCode.SEMANTIC_LITERAL_INT_OVERFLOW,
+                    errorContext = errorContextFrom(metas),
+                    internal = false
+                )
+            }
+        }
+
+    override fun transformTypeDecimalType(node: PartiqlAst.Type.DecimalType): PartiqlLogical.Type.DecimalType =
+        (super.transformTypeDecimalType(node) as PartiqlLogical.Type.DecimalType).apply { // Validation
+            validateDecimalOrNumericType(scale, precision, metas)
+        }
+
+    override fun transformTypeNumericType(node: PartiqlAst.Type.NumericType): PartiqlLogical.Type.NumericType =
+        (super.transformTypeNumericType(node) as PartiqlLogical.Type.NumericType).apply { // Validation
+            validateDecimalOrNumericType(scale, precision, metas)
+        }
+
+    private fun validateDecimalOrNumericType(scale: LongPrimitive?, precision: LongPrimitive?, metas: MetaContainer) {
+        if (scale != null && precision != null && typedOpBehavior == TypedOpBehavior.HONOR_PARAMETERS) {
+            if (scale.value !in 0..precision.value) {
+                err(
+                    "Scale ${scale.value} should be between 0 and precision ${precision.value}",
+                    errorCode = ErrorCode.SEMANTIC_INVALID_DECIMAL_ARGUMENTS,
+                    errorContext = errorContextFrom(metas),
+                    internal = false
+                )
+            }
+        }
+    }
+
+    override fun transformExprCallAgg(node: PartiqlAst.Expr.CallAgg): PartiqlLogical.Expr =
+        super.transformExprCallAgg(node).apply { // Validation
+            val setQuantifier = node.setq
+            val metas = node.metas
+            if (setQuantifier is PartiqlAst.SetQuantifier.Distinct && metas.containsKey(IsCountStarMeta.TAG)) {
+                err(
+                    "COUNT(DISTINCT *) is not supported",
+                    ErrorCode.EVALUATOR_COUNT_DISTINCT_STAR,
+                    errorContextFrom(metas),
+                    internal = false
+                )
+            }
+        }
 }
```

Comment:
Validation should be its own pass using Visitor. It's not a transform

Reply:
Will apply this comment